src/ServiceInsight.Tests/XamlIconsTests.cs

```diff
@@ -0,0 +1,170 @@
+            t.Start();
+            wait.Wait();
+            return results;
+        }
+
+        private static void PrepareForRender(UIElement control)
+        {
+            // From http://stackoverflow.com/a/2596035
+            control.Measure(new Size(double.PositiveInfinity, double.PositiveInfinity));
+            control.Arrange(new Rect(control.DesiredSize));
+
+            control.Dispatcher.Invoke(() => { }, DispatcherPriority.Loaded);
+        }
+
+        public static IEnumerable GetControlTemplates => GetTestCases<ControlTemplate>();
+
+        public static IEnumerable GetDrawingImages => GetTestCases<DrawingImage>();
+
+        public static IEnumerable GetGeometries => GetTestCases<Geometry>();
+
+        [Test]
+        [Explicit]
+        [TestCaseSource(typeof(XamlIconsTests), nameof(GetControlTemplates))]
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public void ControlTemplateTest(string resourceName)
+        {
+            var icon = new ContentControl { Foreground = Brushes.Green };
+            var control = CreateDisplayFrame(resourceName, icon);
+            icon.Template = (ControlTemplate)control.FindResource(resourceName);
+
+            PrepareForRender(control);
+
+            NamerFactory.Clear();
+            using (ApprovalResults.ForScenario(resourceName))
+            {
+                Approvals.Verify(new ImageWriter(f => WpfUtils.ScreenCapture(control, f)));
+            }
+        }
+
+        [Test]
+        [Explicit]
+        [TestCaseSource(typeof(XamlIconsTests), nameof(GetDrawingImages))]
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public void DrawingImagesTest(string resourceName)
+        {
+            var icon = new Image();
+            var control = CreateDisplayFrame(resourceName, icon);
+            icon.Source = (DrawingImage)control.FindResource(resourceName);
+
+            PrepareForRender(control);
+
+            NamerFactory.Clear();
+            using (ApprovalResults.ForScenario(resourceName))
+            {
+                Approvals.Verify(new ImageWriter(f => WpfUtils.ScreenCapture(control, f)));
+            }
+        }
+
+        [Test]
+        [Explicit]
+        [TestCaseSource(typeof(XamlIconsTests), nameof(GetGeometries))]
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public void GeometriesTest(string resourceName)
+        {
+            var icon = new Path { Fill = Brushes.Black, Stretch = Stretch.UniformToFill };
+            var control = CreateDisplayFrame(resourceName, icon);
+            icon.Data = (Geometry)control.FindResource(resourceName);
+
+            PrepareForRender(control);
+
+            NamerFactory.Clear();
+            using (ApprovalResults.ForScenario(resourceName))
+            {
+                Approvals.Verify(new ImageWriter(f => WpfUtils.ScreenCapture(control, f)));
+            }
+        }
+
+        [Test]
+        public void OnlyExpectedResourcesInDictionary()
```

Comment:
The icon file is a resource dictionary, which can contain any type of resource, not just the 3 we are testing for (ControlTemplate, DrawingImage, Geometry). This test ensures that only resources of those 3 types are in the dictionary. That way if we add another way to draw icons, this test will fail and we know to update the icon tests.